### PR TITLE
Swaps deeplink

### PR DESCRIPTION
**Pull request by greg-schrammel**
Fixes APP-1234

## What changed (plus any additional context for devs)

```js
inputAsset: UniqueId
outputAsset: UniqueId

inputAmount: Number
outputAmount: Number
percentageToSell: Number // value between 0 and 1

gasSpeed: 'normal' | 'fast' | 'urgent'
slippage: Number
flashbots: Boolean

from: Address
```

with assets
- `inputAsset` and `outputAsset` expects a `UniqueId (address_chainId)`
ex: `rainbow://swaps?inputAsset=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48_1`

with prefilled values
- `percentageToSell` > `inputAmount` > `outputAmount` only one is used, in this order of importance
- `percentageToSell` expects a value between `0` and `1`
ex: `rainbow://swaps?inputAsset=eth_1&percentageToSell=0.2`, 20% of wallet swappable ETH

with `gasSpeed`
- values are `normal`,`fast` or `urgent` 
ex: `rainbow://swaps?gasSpeed=urgent`

with `flashbots`
- param not included does not mean false, keeps user preferences
- `flashbots=true` any other value besides `true` means `false`
ex: `rainbow://swaps?flashbots=true`

with `slipagge`
- in bips `slippage=100` = 1% slippage
ex: `rainbow://swaps?slippage=100`

with `from`
- let you navigate to another wallet, param is silently ignored if you don't own the wallet 
ex: `rainbow://swaps?from=0x507F0daA42b215273B8a063B092ff3b6d27767aF`

## What to test

- multiple weird combinations of params
- with app open and app closed

to test in ios you can run `xcrun simctl openurl booted 'rainbow://swap/?outputAsset=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48_1&gasSpeed=urgent&flashbots=true&inputAmount=1&inputAsset=eth_1'`

### CODE REVIEW

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/afa7d2ecb7bd8c91a1f526c9c4c4fad7c61232e6.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/afa7d2ecb7bd8c91a1f526c9c4c4fad7c61232e6.ipa) for afa7d2ecb7bd8c91a1f526c9c4c4fad7c61232e6

File: src/handlers/deeplinks.ts [hunk 1]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
     store.dispatch(walletConnectSetPendingRedirect());
   }
 }
+
+const isValidAsset = (s: string | undefined): s is UniqueId => {
+  if (!s) return false;
+  const [address, chainId] = s.split('_');
+  return supportedSwapChainIds.includes(parseInt(chainId, 10)) && (address === 'eth' || address.length === 42);
+};
```
Comment (walmat) [inline src/handlers/deeplinks.ts:362]: what do you think about just deriving the `address` and `chainId` in this function? You're basically checking the same thing after. This function could:
1. check if address and chainId are present and valid according to our `supportedSwapChainIds`, and then return the `address` and `chainId` to be used later in the `querySearchAsset ` function.

File: src/handlers/deeplinks.ts [hunk 2]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+
+  const searchAsset = await queryTokenSearch({
+    chainId,
+    query: address.toLowerCase(),
+    keys: ['address'],
+    threshold: 'CASE_SENSITIVE_EQUAL',
+    list: 'verifiedAssets',
+  }).then(res => res[0]);
+
+  // @ts-expect-error aaaa
```
Comment (walmat) [inline src/handlers/deeplinks.ts:379]: what's the typescript error here?

File: src/handlers/deeplinks.ts [hunk 3]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+
+async function handleSwapsDeeplink(url: string) {
+  const { query } = new URL(url, true);
+
+  setFromWallet(query.from);
+
+  const params: SwapsParams = {};
+
+  const inputAsset = querySearchAsset(query.inputAsset);
+  const outputAsset = querySearchAsset(query.outputAsset);
```
Comment (walmat) [inline src/handlers/deeplinks.ts:409]: possible to do these in parallel?

File: src/logger/index.ts [hunk 4]
```diff
@@ -209,7 +209,7 @@ export class Logger {
   }
 
   debug(message: string, metadata: Metadata = {}, context?: string) {
-    if (context && !this.debugContextRegexes.find(reg => reg.test(context))) return;
+    // if (context && !this.debugContextRegexes.find(reg => reg.test(context))) return;
```
Comment (walmat) [inline src/logger/index.ts:212]: put this back

File: src/utils/branch.ts [hunk 5]
```diff
@@ -7,7 +7,9 @@ import { analyticsV2 } from '@/analytics';
 import * as ls from '@/storage';
 import { logger, RainbowError } from '@/logger';
 
-export const branchListener = async (handleOpenLinkingURL: (url: any) => void) => {
+const isEmpty = <T extends object>(obj: T | undefined): obj is undefined => !obj || Object.keys(obj).length === 0;
```
Comment (walmat) [inline src/utils/branch.ts:10]: could probably just use lodash isEmpty since we have that available.

File: src/utils/branch.ts [hunk 6]
```diff
@@ -27,10 +29,10 @@ export const branchListener = async (handleOpenLinkingURL: (url: any) => void) =
 
     logger.debug(`[branchListener]: handling event`, { params, uri }, logger.DebugContext.deeplinks);
 
-    if (!params && uri) {
+    if (isEmpty(params) && uri) {
```
Comment (walmat) [inline src/utils/branch.ts:32]: good catch!

File: src/handlers/deeplinks.ts [hunk 7]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+  params.inputAsset = await inputAsset;
+  params.outputAsset = await outputAsset;
+
+  if ('flashbots' in query) params.flashbots = query.flashbots === 'true';
+  if ('slippage' in query) params.slippage = query.slippage;
+
+  if (query.percentageToSell) params.percentageToSell = clamp(+query.percentageToSell, 0, 1);
+  else if (query.inputAmount) params.inputAmount = query.inputAmount;
+  else if (query.outputAmount) params.outputAmount = query.outputAmount;
```
Comment (walmat) [inline src/handlers/deeplinks.ts:418]: do we need BigInt checks here?

File: src/handlers/deeplinks.ts [hunk 8]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+    keys: ['address'],
+    threshold: 'CASE_SENSITIVE_EQUAL',
+    list: 'verifiedAssets',
+  }).then(res => res[0]);
+
+  // @ts-expect-error aaaa
+  if (!searchAsset) return userAsset;
+
+  return parseSearchAsset({ searchAsset, userAsset: await userAsset });
+};
```
Comment (walmat) [inline src/handlers/deeplinks.ts:383]: I think a lot of this can be done in parallel so we can speed up this quite a bit.

File: src/state/swaps/swapsStore.ts [hunk 9]
```diff
@@ -27,6 +27,9 @@ export interface SwapsState {
   selectedOutputChainId: ChainId;
   outputSearchQuery: string;
 
+  percentageToSell: number; // Value between 0 and 1, e.g., 0.5, 0.1, 0.25
```
Comment (walmat) [inline src/state/swaps/swapsStore.ts:30]: when a user passes in a `percentageToSell ` like `0.123131231321321` should we convert it to `0.12` and so on?

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 10]
```diff
@@ -563,6 +593,10 @@ export function useSwapInputsController({
     autoStart: false,
   });
 
+  if (internalSelectedInputAsset.value && internalSelectedOutputAsset.value) {
+    fetchQuoteAndAssetPrices();
+  }
+
```
Comment (walmat) [inline src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:599]: i don't think this is necessary since it should be handled via a `useAnimatedReaction`.

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 11]
```diff
@@ -657,6 +691,7 @@ export function useSwapInputsController({
       assetToSellNetwork: internalSelectedInputAsset.value?.chainId,
     }),
     (current, previous) => {
+      if (!previous) return;
```
Comment (walmat) [inline src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:694]: why was this added?

File: src/__swaps__/screens/Swap/providers/swap-provider.tsx [hunk 12]
```diff
@@ -132,28 +133,48 @@ interface SwapProviderProps {
...
 
+const getInitialSliderXPosition = ({
+  inputAmount,
+  maxSwappableAmount,
+}: {
+  inputAmount: string | undefined;
+  maxSwappableAmount: string | undefined;
+}) => {
+  if (inputAmount && maxSwappableAmount) {
+    return clamp(+mulWorklet(divWorklet(inputAmount, maxSwappableAmount), SLIDER_WIDTH), 0, SLIDER_WIDTH);
```
Comment (walmat) [inline src/__swaps__/screens/Swap/providers/swap-provider.tsx:144]: will `maxSwappableAmount` always be defined here? Could there be a case where it's undefined as declared in the types? Division by 0?

**Review by walmat — COMMENTED** [review]
Mostly comments that I need clarity on, a few actual changes that I think could make the flow even better:
1. parallelize some of the network requests
2. safeguards against division by 0, undefined cases, etc.
3. make sure the default 50% slider position if no inputAmount is defined hasn't regressed.

Otherwise, great work!

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/d51e895db95b897058e4f4132926fb36e56b080a.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/d51e895db95b897058e4f4132926fb36e56b080a.ipa) for d51e895db95b897058e4f4132926fb36e56b080a

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts [hunk 13]
```diff
@@ -50,10 +50,37 @@ import { deepEqualWorklet } from '@/worklets/comparisons';
...
+  sliderXPosition: number;
+}) {
+  if (inputAsset && inputAmount) {
+    const inputNativeValue = mulWorklet(inputAmount, inputAsset?.price?.value ?? 0);
+    return { inputAmount, inputNativeValue, outputAmount: 0, outputNativeValue: 0 };
+  }
+  if (outputAsset && outputAmount) {
+    const outputNativeValue = mulWorklet(outputAmount, outputAsset?.price?.value ?? 0);
+    return { inputAmount: 0, inputNativeValue: 0, outputAmount, outputNativeValue };
+  }
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:75]: the idea was to prefer `inputAmount > outputAmount > percentageToSell`
can never be true at the same time because if we have `inputAmount` output is calculated from that and vice versa 
it correctly sets the inputAmount to 50% if not especified, that happens in `getInputValuesForSliderPositionWorklet `

File: src/__swaps__/screens/Swap/providers/swap-provider.tsx (hunk 12 above)
Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/providers/swap-provider.tsx:144]: that `if` should check for that

File: src/handlers/deeplinks.ts [hunk 14]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+  const [address, chainId] = s.split('_');
+  return supportedSwapChainIds.includes(parseInt(chainId, 10)) && (address === 'eth' || address.length === 42);
+};
+
+const querySearchAsset = async (uniqueId: string | undefined): Promise<ParsedSearchAsset | undefined> => {
+  if (!isValidAsset(uniqueId)) return;
+
+  const { address, chainId } = deriveAddressAndChainWithUniqueId(uniqueId);
+
+  const userAsset = queryUserAssets().then(allUserAssets => allUserAssets[chainId][uniqueId]);
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts:358]: so it's non blocking, it only `await` the `userAsset` when needed later
it's gonna send this request move to the `searchAsset` one which is awaited/blocking, and later hopefully we will already have this `userAsset` to return with it
would be weird to do the `allUserAssets[chainId][uniqueId]` without extracting to a new function 

wdyt?

`Promise.all` would look the same, something like
```ts
const [userAsset, searchAsset] = Promise.all([
  queryUserAssets().then(allUserAssets => allUserAssets[chainId][uniqueId]),
  queryTokenSearch({ 
    ... 
  })
])
```

[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts:358]: ah this works lol, I can do this if you think it's clearer don't mind either way
```ts
const [allUserAssets, searchAsset] = Promise.all([  
  queryUserAssets(),
  queryTokenSearch({ 
    ... 
  })
])
const userAsset =  allUserAssets[chainId][uniqueId]
```

File: src/handlers/deeplinks.ts [hunk 15]
```diff
@@ -337,3 +354,71 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+function setFromWallet(address: string | undefined) {
+  if (!address || !isAddress(address)) return;
+
+  const userWallets = store.getState().wallets.wallets!;
+  const userAddresses = Object.values(userWallets).flatMap(w => w.addresses.map(a => a.address));
+
+  if (userAddresses.includes(address)) {
+    store.dispatch(settingsUpdateAccountAddress(address));
+  }
+}
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts:388]: I'm not sure what to do here

File: src/handlers/deeplinks.ts (hunk 7 above)
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts:407]: added number validations

File: src/logger/index.ts (hunk 4 above)
Comment (greg-schrammel) [reply src/logger/index.ts:212]: ops

File: src/state/swaps/swapsStore.ts (hunk 9 above)
Comment (greg-schrammel) [reply src/state/swaps/swapsStore.ts:30]: should we? I don't see percentage to sell as something to worry about precision, but shouldn't convert to `0.12` rn

File: src/utils/branch.ts (hunk 5 above)
Comment (greg-schrammel) [reply src/utils/branch.ts:10]: I don't know lodash so went to look at what it does, I don't think we need all that, and it "breaks" the types
https://github.com/lodash/lodash/blob/4.17.15/lodash.js#L11479

File: src/handlers/deeplinks.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts]: ok found out I could replace that `queryUserAssets` with `userAssetsStore.getState().getUserAsset(uniqueId)` which already has the right type uhu!

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/7e95ffbd39ff08c549ccd82ad48884b5328884f3.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/7e95ffbd39ff08c549ccd82ad48884b5328884f3.ipa) for 7e95ffbd39ff08c549ccd82ad48884b5328884f3

File: src/state/swaps/swapsStore.ts (hunk 9 above)
Comment (walmat) [reply src/state/swaps/swapsStore.ts:30]: just a question, up to you whether you deem it necessary. I personally think it's fine.

File: src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts (hunk 11 above)
Comment (walmat) [reply src/__swaps__/screens/Swap/hooks/useSwapInputsController.ts:696]: still curious about this

**Comment by derHowie** [comment]
hey @greg-schrammel, did we end up getting blocked / deprioritizing this?

**Comment by greg-schrammel** [comment]
> hey @greg-schrammel, did we end up getting blocked / deprioritizing this?

no I just moved to other stuff and forgot about it, just fixed the conflicts

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/4c3d7cf66f3369354fb8b3db27d0dd510ea1a3d0.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/4c3d7cf66f3369354fb8b3db27d0dd510ea1a3d0.ipa) for 4c3d7cf66f3369354fb8b3db27d0dd510ea1a3d0

**Review by walmat — APPROVED** [review]

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/dc3803c278866e39eff1afbac08477465b2ce92a.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/dc3803c278866e39eff1afbac08477465b2ce92a.ipa) for dc3803c278866e39eff1afbac08477465b2ce92a

File: src/handlers/deeplinks.ts [hunk 16]
```diff
@@ -326,3 +343,72 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+
+  const inputAsset = querySwapAsset(query.inputAsset);
+  const outputAsset = querySwapAsset(query.outputAsset);
+
+  if ('flashbots' in query) params.flashbots = query.flashbots === 'true';
+  if ('slippage' in query && isNumbericString(query.slippage)) params.slippage = query.slippage;
+
+  if (isNumbericString(query.percentageToSell)) params.percentageToSell = clamp(+query.percentageToSell, 0, 1);
+  else if (isNumbericString(query.inputAmount)) params.inputAmount = query.inputAmount;
+  else if (isNumbericString(query.outputAmount)) params.outputAmount = query.outputAmount;
```
Comment (derHowie) [inline src/handlers/deeplinks.ts:405]: I don't want to make a style rule around this because it'd be annoying to enforce, but this is more difficult to read w/o brackets imo.

File: src/__swaps__/screens/Swap/resources/assets/userAssets.ts [hunk 17]
```diff
@@ -70,9 +71,22 @@ type UserAssetsQueryKey = ReturnType<typeof userAssetsQueryKey>;
...
+  currency = store.getState().settings.nativeCurrency,
+  testnetMode = false,
+}: Partial<FetchUserAssetsArgs> = {}) {
+  const queryKey = userAssetsQueryKey({ address, currency, testnetMode });
+
+  const cachedData = queryClient.getQueryData<ParsedAssetsDictByChain>(queryKey);
+  if (cachedData) return cachedData;
+
+  return userAssetsFetchQuery({ address, currency, testnetMode });
+}
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/resources/assets/userAssets.ts:88]: I don't think so, but we already have some with "fetch" prefix that doesn't account for the cache like this one, I agree the name may not be the best but can't really think of a better one rn, lmk if you have an idea

File: src/handlers/deeplinks.ts [hunk 18]
```diff
@@ -326,3 +343,72 @@ function handleWalletConnect(uri?: string, connector?: string) {
...
+
+  if (!wallet) return;
+
+  await Promise.all([store.dispatch(walletsSetSelected(wallet)), store.dispatch(addressSetSelected(address))]);
+}
+
+function isNumbericString(value: string | undefined): value is string {
+  if (!value) return false;
+  return !isNaN(+value);
+}
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/handlers/deeplinks.ts:388]: ops

**Review by derHowie — APPROVED** [review]

File: src/__swaps__/screens/Swap/navigateToSwaps.ts [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  const state = useSwapsStore.getState();
+  return {
+    inputMethod,
+    lastTypedInput,
+    focusedInput: lastTypedInput,
+    inputAsset: params.inputAsset || state.inputAsset,
+    outputAsset: params.outputAsset || state.outputAsset,
+    ...params,
+  } as const;
+}
```
[the remark this answers is not in the document]
Comment (greg-schrammel) [reply src/__swaps__/screens/Swap/navigateToSwaps.ts:55]: they are typed as `ParsedSearchAsset | null`

**Comment by brunobar79** [comment]
Launch in [simulator](http://localhost:29070/install/ios?virtual=https://app-team.p.rainbow.me/swaps-deeplink/60f5effece4522f2c53de1d7bcd0db84d165fb8b.app.zip) or [device](http://localhost:29070/install/ios?physical=https://app-team.p.rainbow.me/swaps-deeplink/60f5effece4522f2c53de1d7bcd0db84d165fb8b.ipa) for 60f5effece4522f2c53de1d7bcd0db84d165fb8b
